### DISCUSSION TITLE
Unusual ingress message rate values reported from GET /api/queues/{name}

### DISCUSSION THREAD

**Discussion by laithalissa** [rabbitmq/rabbitmq-server#8419] [Other]
### Describe the bug

I'm experiencing some alarming inconsistencies with the queue message rate metrics from the management API which I'm depending on very heavily.

I'm finding that rate is occasionally reported as an incorrect value, orders of magnitude more than it really is.

To demonstrate, I submitted 3 messages, separated by a few seconds, to an exchange to which a queue is subscribed.
I have a while loop hitting the endpoint to monitor the publish stats over this time period.

```bash
http://localhost:15672/api/queues/%2F/myQueue' \
| jq '.message_stats.publish_details'
```
This is the output- I've removed the duplicate output to keep it readable

```
2023-05-30 16:35:35 {"rate":0}
  // [21s] No change in this time period
2023-05-30 16:35:56 {"rate":0}
2023-05-30 16:35:56 {"rate":0.2}
  // [ 8s] No change in this time period
2023-05-30 16:36:04 {"rate":0.2}
2023-05-30 16:36:04 {"rate":0}
  // [ 8s] No change in this time period
2023-05-30 16:36:12 {"rate":0}
2023-05-30 16:36:12 {"rate":740}
  // [ 7s ] No change in this time period
2023-05-30 16:36:19 {"rate":740}
2023-05-30 16:36:20 {"rate":0}
```

For the first message, the rate didn't change, it remained zero
For the second message, I saw an 8 second period of rate: 0.2
For the third message, I saw a 7 second period of rate: 740

Here is a screencap of the rabbit ui for this queue, capturing this activity.
![Screenshot 2023-05-30 at 17 37 47](https://github.com/rabbitmq/rabbitmq-server/assets/887720/f31bccf2-9c70-4eeb-bdd2-4ae9a7a2043d)

RabbitMQ 3.11.15
Erlang 25.3.2
Deployed on kubernetes using the bitnami rabbit-cluster-operator, version, chart version 3.3.2
Exchange is of topic type, queue is durable, quorum type. There's nothing else using the exchange or queue when I test.

### Reproduction steps

This issue doesn't reproduce consistently 

Assuming the setup in my description:
1. Create exchange of type topic
2. Create durable quorum queue, add binding to topic
3. Monitor the `/api/queues/%2F/myQueue` endpoint for metrics (.message_stats.publish_details)
4. Submit a single message to the queue
5. Repeat many times until you see the `rate` spike to an usually high value

### Expected behavior

`rate` should be a value which reasonably reflect the rate of messages

### Additional context

I've reproduced this issue about 7 times (I ran approx 50 tests), here are the unique rate values I yielded, the 0.4, 0.6 values had 2-3 messages  submitted at once, respectively, so they can be ignored

"rate":0
"rate":0.2
"rate":0.4
"rate":0.6
"rate":730.4
"rate":731
"rate":731.2
"rate":734.8
"rate":740
"rate":743.4

**Comment by michaelklishin** [comment]
I'm afraid that's not enough information to suggest anything specific. The commended monitoring solution is [Prometheus](https://rabbitmq.com/prometheus.html) (or a compatible tool), to the point
that management UI long has had a way to disable metrics entirely.

Metric samples are emitted periodically. HTTP API then aggregates them across all nodes.

The chart does not contain any values that would be in the 700+ range. My best guess is that some
sliding window calculations smoothen the irregularity.

Assuming that you use this for backlog monitoring, a single reported value is not going to be useful in practice. It would make you depend too much on the specific moment the value is retrieved.
Given that the rates are computed over a sliding window of samples, they are never perfectly accurate.

**Reply by laithalissa** [reply]
I'm using Keda to scale based on message inflow- it's a task queue, but it needs to be as responsive as possible. In order to monitor rabbitmq based on message rate, keda supports only using the http api (see https://keda.sh/docs/2.10/scalers/rabbitmq-queue/).

Prometheus is a poor choice for my use case, because I want to minimize scaling response time as much as possible, and prometheus will add a scrape interval to the delay. I do run prometheus- but I use it only for observability- I didn't like the idea of abusing prometheus's global scrape interval simply to get a higer resoltion view of the rabbit queue status.

> The chart does not contain any values that would be in the 700+ range

I'm not sure what this means.

I'm aware that the instant metric value isn't ideal, but it's all Keda seems to support out the box.

I previously raised this [issue](https://github.com/kedacore/keda/issues/4507), do you think it's sensible to provide a solution for my use-case?

**Reply by michaelklishin** [reply]
I agree with the idea in the issue. It's surprising that an autoscaling tool would make such a mistake: depend on a single value and not smoothen (or at least eliminate outliers) the rate over a period of time.

But to be fair, VMware's own monitoring tools got every possible thing wrong at first, and RabbitMQ
often was the tool that had exposed the flaws.

Since Keda is a Kubernetes-oriented tool, I am also surprised that Prometheus is not the source of the data.

**Reply by laithalissa** [reply]
I'm AFK at the moment so can't verify, but I seem to remember not being able to interrogate rabbitmq's prom metrics to get the rate of messages published to a specific queue, iirc it was only available for exchanges. I'll check when I'm back at my keyboard and update.

**Reply by michaelklishin** [reply]
[Aggregate vs. Per-object Metrics](https://rabbitmq.com/prometheus.html#metric-aggregation)

**Reply by laithalissa** [reply]
Sorry for the delay @michaelklishin , brain ran out of awake-credits.

I've checked the details again, and as far as I understand it, there does seem to be a functional limitation using prometheus metrics in this case.

The detailed metrics allow us to monitor a queue with `queue_coarse_metrics` :
<img width="1021" alt="image" src="https://github.com/rabbitmq/rabbitmq-server/assets/887720/76b6187e-ead9-455e-8b7b-3dd51bfacef8">

However this only gives us information about the messages currently in the queue. Hence unless you have 0 consumers removing message, you won't be able to derive a publish rate.

You also have the `channel_queue_exchange_metrics` family

<img width="988" alt="image" src="https://github.com/rabbitmq/rabbitmq-server/assets/887720/806ed6a6-94dc-4e62-94a2-a260cc1fd0be">

The issue with leveraging this is it's not trivial to get the rate of messages published to a specific queue. The reason for this is that the metric includes the channel as a label (as an aside this is expensive for prom due to the high cardinality), so to get a somewhat-useful metric to monitor, you need to either create a recording rule or use a subquery, as I have here:

[Edit: Corrected graph]
<img width="1117" alt="image" src="https://github.com/rabbitmq/rabbitmq-server/assets/887720/31cf8b87-dd32-4dbe-93db-7fed8a0b91c8">

Note that this is with a 2s scrape interval, submitting a message every 0.5s (request takes about 0.4s to complete), so call it approx 1 publish/second

IMHO this graph produces less smooth results than the HTTP api. If not for the spikey values I mentioned in the OP, I would consider it the better option here.

I'm really in a bind here. I feel like neither option is going to deliver a great solution.

Do you think it's worth implementing the HTTP rate smoothing as I proposed in the other ticket to solve my issue, or do you have a better idea?

**Reply by Zerpet** [reply]
I'm not sure what's your use case, as in, do you want to scale producers if publish rate is low? Do you want to scale consumers if publish rate is high?

I'm going to gues that you want to scale out consumers. I think that basing the scaling rule on published message rate may not be the best solution. The published rate only tells one side of the story, and it is not enough to make a decision about scaling IMO. For example, you may have e.g. 500 msg/s published in a queue, if you have 1 consumer capable of keeping up with the publish rate (so that messages ready stays near zero), there is no need to scale out consumers.

However, if you observe the number of ready messages grow over time, it is a clear sign that consumers are not keeping up. In this case, it makes sense to scale out consumers, or de-scale producers.

If you know, on average, how many ready messages are acceptable, you could base your scaling rule for consumers on number of ready messages.

KEDA could use RabbitMQ Prometheus endpoint (instead of Prometheus) to fetch the metrics it needs, but there is some filtering and discarding required, since, as you mentioned, there doesn't seem to be a way to filter our queues from the detailed prometheus endpoint.

**Reply by laithalissa** [reply]
@Zerpet to give you a bit of context, this queue is effectively a buffer for real time requests. It takes about 5-8s to process a request, but when we get an unexpected increase of messages, we're unable to keep up with our target "sla" of under 10s. Scaling workers is a _very_ slow process, it can take 3-5 minutes for a consumer to be ready after a scale event. As such I want to scale at the earliest possible time. I'm attempting to implement this with 2 keda scalers. One to trigger a scale for when the message rate exceeds a threshold, and a second (backup) scaler which scales when there's a backlog of message. The second scaler is an emergency response for when we're exceeding our anticipated growth (or something changes regarding processing time). Due to the unavoidable initialisation times, I'm not able to accomodate any delay in the scaling responses, hence I'm always aiming for `0` ready messages, and am very reluctant to add a potentially large scrapeInterval delay in my solution.

In order to make it work with prometheus, I would need to reconfgiure prometheus with a sufficiently low enough global scrape interval, in the order of a few seconds, which isn't ideal. Granted it's not much different than having Keda repeatidly poll the rabbitmq http api, but I feel it's compromising my general-purpose monitoring tool just to workaround this issue, since prometheus is the subject to managing much higher resolution data, and I have to be careful regarding other service monitors going forward to ensure they don't flood prometheus with low interval metrics.

The options as I see them are:
1. Suck it up and use prometheus with a very low scraping interval
2. Run a second prometheus as a dedidcated shim between rabbit and keda
3. Write a proxy to handle this for me (keda external scaler)
4. Implement/wait for Keda team to implement a smarter http call to rabbit

@michaelklishin you put a negative spin on the HTTP api previously, is there a plan/thought to deprecate and remove it in favour of the promtheus metrics endpoint?

**Reply by michaelklishin** [reply]
I only meant that for *metrics*, there is a better option: Prometheus. It's been possible to [disable metrics in the HTTP API](https://rabbitmq.com/management.html#disable-stats) for several years.

With everything else that the HTTP API provides, it does a good job, and there are no plans to deprecate the entire API. It does not seem to be realistic for us to remove the metrics part entirely, we cannot require every user to always have a Prometheus installation around.

For production systems, the recommendation is to use Prometheus for a variety of reasons. Long term
metric storage is one good example.

**Reply by Zerpet** [reply]
Thanks for the context! It's an interesting use case. I agree with the issue you reported in Keda https://github.com/kedacore/keda/issues/4507. If that doesn't happen soon-ish, you could try to [amqpprox](https://github.com/bloomberg/amqpprox) by Bloomberg for a proxy and Keda external scaler (option 3). Option 2 also sounds good to me, in combination with a Prometheus Scaler, could do the trick.

Regarding the inaccurate metric coming from the HTTP API, our focus is on the Prometheus endpoint. That being said, we'll definitely welcome a PR, although I think it'll be best to invest on https://github.com/kedacore/keda/issues/4507

A long shot, have you tried with `3.12.0-rc.4` as Michael tried out? DockerHub should already have the RC images built.

**Reply by laithalissa** [reply]
@Zerpet Thanks for the suggestions. 

I haven't tried `3.12.0-rc.4` yet, I didn't want to introduce unknowns from an RC, but maybe I should since you're scheduled for a stable release soon. Trouble is it's not that easy to reproduce. I had been using this version for ~4 weeks when I suddenly got alerted that the cloud provider span up a 400 instances for one request 💀 . Fun times.

@michaelklishin thank you for clarifying.

I'll followup with any findings from `3.12.0-rc.4`.

**Reply by laithalissa** [reply]
Reproduced the issue on `3.12.0-rc.4`. The rate isn't as high as it was previously.

I found a sweet-spot of leaving 5 minutes between requests (`while true; do qmsg; sleep 300; done;`), yielding `{"rate":47.2}` and `{"rate":46}`, so far.

**Comment by michaelklishin** [comment]
Given a quorum queue named `qq.1` on a 3.12.0-rc.4, I publish a message per second using Bunny in a loop:

``` ruby
c = Bunny.new; c.start
ch = c.create_channel
qq = ch.quorum_queue("qq.1")

loop { qq.publish(""); sleep 1 }
```

and I monitor the rate using

``` shell
for i in {1..1000}; do curl -X GET -s -u guest:guest http://localhost:15672/api/queues/%2F/qq.1 | jq ".message_stats.publish_details"; sleep 1; done
```

The output is

```
null
null
null
null
null
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0.6
}
{
  "rate": 0.6
}
{
  "rate": 0.6
}
{
  "rate": 0.6
}
{
  "rate": 0.6
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0
}
{
  "rate": 0.2
}
{
  "rate": 0.2
}
{
  "rate": 0.2
}
{
  "rate": 0.2
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
{
  "rate": 1
}
```

**Comment by michaelklishin** [comment]
When I change the publishing rate to be about 5 messages per second

``` ruby
loop { qq.publish(""); sleep 0.2 }
```

I get the following:

```
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 4.8
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
{
  "rate": 5
}
```

**Comment by mkuratczyk** [comment]
I've also tried with different message rates, different number of publishers and some other modifications and never seen a value higher than expected.

**Reply by laithalissa** [reply]
Interesting, I've reproduced it in 2 different versions of rabbitmq in a production environment. I can maybe try to repro in a local setup with docker compose or something. What testing setup are you using for this? Bear in mind I need to simulate an HA rabbit cluster setup with a quorem queue
